### PR TITLE
CI

### PR DESCRIPTION
**Pull request by tombentley**
I think this is actually more correct dependency-wise, since
the output of the mvn package is needed in the docker image.

This change also means the CI will run the Java tests, since the
.travis/build.sh does a docker_build.

### CODE REVIEW

**Comment by tombentley** [comment]
Before merging this, I need either to fix #174, or disable the `ControllerIntegrationTest` when running on CI.

**Comment by tombentley** [comment]
This now fixes #174, but since getting the test working with minikube meant a few changes, @scholzj could you review again.

**Comment by ppatierno** [comment]
@tombentley I'd like to review this one but I can't do that right now. Can you wait to merge until tomorrow morning please ? Thanks !

**Comment by tombentley** [comment]
Of course!

File: topic-controller/src/test/java/io/strimzi/test/KubeCluster.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            }
+        } catch (IOException e) {
+            throw new KubeClusterException(e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new KubeClusterException(e);
+        }
+    }
+
+    protected String exec2(String... cmd) throws KubeClusterException {
```
Comment (ppatierno) [inline topic-controller/src/test/java/io/strimzi/test/KubeCluster.java]: can we define a common `exec` method with a `boolean redirectOutput` so that if that's true it goes creating the tmp file redirecting output there ? We could avoid to have `exec` and `exec2` in this way. Wdyt ?

File: topic-controller/src/test/java/io/strimzi/test/Oc.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ */
+
+package io.strimzi.test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class Oc extends KubeCluster {
```
Comment (ppatierno) [inline topic-controller/src/test/java/io/strimzi/test/Oc.java]: I have a doubt about this `Oc` name. We are going to spin an OpenShift cluster using `oc` tool. The other class is called `Minikube`, not `Kubectl` :-) Maybe we could have another class named `Minishift` if we need to spin the OpenShift cluster in that way but I would call this class just `OpenShiftCluster` ... but it's just my opinion :-)

File: .travis/setup-kubernetes.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+curl -Lo minikube https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64 && chmod +x minikube
+curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && chmod +x kubectl
+
+mkdir -p /tmp/openshift
+wget https://github.com/openshift/origin/releases/download/v3.7.0/openshift-origin-client-tools-v3.7.0-7ed6862-linux-64bit.tar.gz -O openshift.tar.gz
+tar xzf openshift.tar.gz -C /tmp/openshift --strip-components 1  
+sudo cp /tmp/openshift/oc /usr/bin
```
Comment (ppatierno) [inline .travis/setup-kubernetes.sh]: If on Travis CI we can't use `oc` tool due to old systemd, why we are pulling the tool here ?

File: .travis/setup-kubernetes.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+curl -Lo minikube https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64 && chmod +x minikube
+curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && chmod +x kubectl
+
+mkdir -p /tmp/openshift
+wget https://github.com/openshift/origin/releases/download/v3.7.0/openshift-origin-client-tools-v3.7.0-7ed6862-linux-64bit.tar.gz -O openshift.tar.gz
```
Comment (scholzj) [inline .travis/setup-kubernetes.sh]: Why are we downloading OpenShift if ti doesn't run on Travis?

File: topic-controller/src/test/java/io/strimzi/test/RoleBinding.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import java.lang.annotation.Target;
+
+/**
+ * Describes a Kubernetes RBAC {@code RoleBinding}.
+ * @see <a href="https://kubernetes.io/docs/admin/authorization/rbac">Kubernetes RBAC docs</a>
+ */
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface RoleBinding {
+    /** The name of the role binding. If not specified it will default to the Role name + "-binding". */
```
Comment (scholzj) [inline topic-controller/src/test/java/io/strimzi/test/RoleBinding.java]: Do we need to sync this with the YAML files / YAML files with this? Cluster Controller YAML files are `strimzi-cluster-controller-role` and ` strimzi-cluster-controller-binding` ... so adding `-binding` to the role name would not work. But I'm not sure how this is related to the YAML files right now.

**Comment by ppatierno** [comment]
@tombentley I have added few comments btw I really like the way to use JUnit rules, external resource and annotations for doing this stuff. Congratulations ! ;)

File: topic-controller/src/main/java/io/strimzi/controller/topic/Config.java [hunk 6]
```diff
@@ -124,6 +125,8 @@ private Value(String key, Type<? extends T> type, boolean required, String doc)
             "A comma-separated list of key=value pairs for selecting ConfigMaps that describe topics.");
     public static final Value<String> KAFKA_BOOTSTRAP_SERVERS = new Value(TC_KAFKA_BOOTSTRAP_SERVERS, STRING, true,
             "A comma-separated list of kafka bootstrap servers.");
+    public static final Value<String> NAMESPACE = new Value(TC_NAMESPACE, STRING,true,
```
[the remark this answers is not in the document]
Comment (tombentley) [reply topic-controller/src/main/java/io/strimzi/controller/topic/Config.java]: touché! ;-)

File: topic-controller/src/test/java/io/strimzi/test/KubeCluster.java (hunk 1 above)
Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/test/KubeCluster.java]: Yeah, I thought about that, but I don't think it would make the code much better, because in the one case you need to mess around with a temp file (`try/finally`) and in the other you don't. I suppose the perf isn't so important here though, so maybe it doesn't matter if we create a temp file and then don't actually use it for anything. Let me think about this.

File: topic-controller/src/test/java/io/strimzi/test/Oc.java (hunk 2 above)
Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/test/Oc.java]: >  The other class is called `Minikube`, not `Kubectl`

It's started using `minikube`, even some of the methods use `kubectl`. I didn't add a `Minishift` because I didn't need one right now. If and when we need a `Minishift` I think there would be an opportunity to refactor because it would have a lot in common with `Minikube`. Probably the same class with static factory methods for each kind.

File: topic-controller/src/test/java/io/strimzi/test/RoleBinding.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    String name() default "";
```
[the remark this answers is not in the document]
Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/test/RoleBinding.java]: Good catch, thanks!

File: .travis/setup-kubernetes.sh (hunk 3 above)
Comment (tombentley) [reply .travis/setup-kubernetes.sh]: I just copied this script from enmasse, since @lulf has done all the hard work of making this work. I think openshift is downloaded because (until he gave up due to systemd issue) he was also trying to make openshift work too. If I remove it here we end up diverging from enmasse. Let's see what @lulf says about fixing this in enmasse.

File: topic-controller/src/test/java/io/strimzi/controller/topic/ControllerIntegrationTest.java [hunk 8]
```diff
@@ -65,22 +66,31 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 
+@RoleBinding(
+        name="strimzi-role-binding",
+        role="strimzi-role",
+        users="developer"
+)
+@Role(
```
[the remark this answers is not in the document]
Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/controller/topic/ControllerIntegrationTest.java]: Something for a future PR.

File: topic-controller/src/test/java/io/strimzi/test/RoleBinding.java (hunk 5 above)
Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/test/RoleBinding.java]: Yeah, I'm not sure this bit of laziness is really worthwhile. We should make the role binding name mandatory in the annotation.

File: topic-controller/src/test/java/io/strimzi/test/Oc.java (hunk 2 above)
Comment (ppatierno) [reply topic-controller/src/test/java/io/strimzi/test/Oc.java]: But what about renaming `Oc` to something like `OpenShiftCluster`. At the end if I think more on that `Oc` is extending `KubeCluster` but oc is a "cluster" itself ... it's the tool. It's a nit I know ... but the oc name for this class isn't clear to me.

Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/test/Oc.java]: Well this is interesting (and perhaps it exposes my ignorance). `oc` is, on the one hand the OpenShift client. But running `oc cluster up` is not the same thing as `minishift start`. But minishift is also OpenShift, isn't it? So can I do a `minishift start` and then use `oc` with that cluster? Presumably I can also use `kubectle` with an OpenShift Cluster? So logically, perhaps I need to separate the starting of a cluster from the tool I use to interact with it once it's running. Conceptually woudl that make more sense? Since both `Oc` and `Minikube` are stateless that wouldn't be difficult to do.

Comment (ppatierno) [reply topic-controller/src/test/java/io/strimzi/test/Oc.java]: `oc` is just the tool. The `cluster up` command just downloads an OpenShift Origin Docker image and starts this image on your local PC in order to have an OpenShift cluster running.
At this point you can use all the other `oc` commands (i.e. get pods, create, ...) for interacting with the OpenShift cluster.
`minishift start` starts a Virtual Machine and inside it does something like the `oc cluster up` does on your local PC do downloading Docker image and starting it : at this point the OpenShift cluster is running in a VM (which is your single cluster node, while in the previous case it's your PC). With `minishift` you have a out-of-box environment and OpenShift cluster without dirtying your PC because all is downloaded and run in a VM.

> Presumably I can also use kubectle with an OpenShift Cluster?

Yes for the common features between Kube and OpenShift

What I'm suggesting is just changing the name to `Oc` into something more related to a "cluster" name. `minikube` is a tool but finally it's a cluster :-) for this reason I like having `Minikube` derived from `KubeCluster` but not `Oc`

Comment (ppatierno) [reply topic-controller/src/test/java/io/strimzi/test/Oc.java]: But as I said it's a nit ... more "naming" related not so important .. I can live with `Oc` ;)

Comment (tombentley) [reply topic-controller/src/test/java/io/strimzi/test/Oc.java]: I have refactored this into `KubeCluster` hierarchy (`OpenShift`, `Minikube` subclasses)  and a `KubeClient` hierarchy (`Oc` and `Kubectl` subclasses) and I think it's nicer. It will allow us to expose the client through the `KubeClusterResource`, so that, for example, individual tests can interact with the cluster in their own way. Thanks for pushing me into doing this @ppatierno!

File: .travis/setup-kubernetes.sh (hunk 3 above)
Comment (tombentley) [reply .travis/setup-kubernetes.sh]: @lulf added `oc` download specifically in [this commit](https://github.com/EnMasseProject/enmasse/commit/04fc947aa68fa08b2da6f64f6d368e2c906df06c), but I don't know what "useful debug info" he needed it for.

Comment (tombentley) [reply .travis/setup-kubernetes.sh]: Apparently they use `oc` for things like `rsh` and `extract`. With `rsh` we could test some interesting corner cases like killing processes etc, so I think we should keep the `oc` download for now. We can always remove it later when it turns out that it's making the ci builds too slow.
